### PR TITLE
feat: support vscode multi-root workspaces

### LINKED ISSUE #2379: Support multi-root workspaces in LSP server
Adopting multi-root workspaces has been on our radar since 2019. The feature was introduced in 2017.  This is a nonstarter for many users who rely on this vscode feature for work!  At this point it's been so long, I would consider the lack of the feature a bug, even though it would have been a feature originally.

The implementation will not be simple.

It will require keeping multiple graphql config instances in memory in `GraphQLLanguageService`!! And combining settings very carefully

https://github.com/microsoft/vscode/wiki/Adopting-Multi-Root-Workspace-APIs

### PR DESCRIPTION
**Pull request by Foo-x**
Closes #2379 

I reffered to the following official implementation.
https://github.com/microsoft/vscode-extension-samples/blob/main/lsp-multi-server-sample/client/src/extension.ts

`graphql-language-service-server` will be started for each folders, and the corresponding schema is used for each.
image

The schema for `packages/app` folder has `foo`.
image
image

The schema for `packages/api` folder does not have `foo`.
image
image

`StatusBarItem` and `OutputChannel` will be shared by all folders and updated according to the active editor.

### CODE REVIEW

**Comment by acao** [comment]
Wow ok, i hadn’t thought of starting multiple instances. Does this work when multiple graphql config projects are defined in a file?

**Comment by Foo-x** [comment]
Sorry, I have not checked on multiple projects.
It is midnight here in Japan now, so please let me check tomorrow.

**Comment by acao** [comment]
@Foo-x hey that would be great, if I don’t get to it first! No hurry at all, but don’t be surprised if this gets released before you wake up!

File: packages/vscode-graphql/src/extension.ts [hunk 1]
```diff
@@ -17,93 +19,134 @@ import {
...
+      debug: {
+        module: serverModule,
+        transport: TransportKind.ipc,
+        options: { ...(debug ? debugOptions : {}) },
+      },
+    };
+
+    const clientOptions: LanguageClientOptions = {
+      documentSelector: [
+        {
```
Comment (acao) [inline packages/vscode-graphql/src/extension.ts:65]: 💯

File: packages/vscode-graphql/src/extension.ts [hunk 2]
```diff
@@ -17,93 +19,134 @@ import {
...
+          scheme: 'file',
+          language: 'graphql',
+          pattern: `${folder.uri.fsPath}/**/*`,
```
Comment (acao) [inline packages/vscode-graphql/src/extension.ts:68]: this is much better, thank you

**Comment by acao** [comment]
@Foo-x i think this is good to go, can you add a changeset file with your changelog entry as well? 

it seems there are unrelated issues with `graphiql-config` multi-project anyways, introduced by me last year, and this should work in tandem with that programatically, my concern is just resource usage, but I will make the optimization to address that concern shortly

**Comment by Foo-x** [comment]
@acao Thank you for adding the changeset!

Regarding the behavior in multi-projects, I found a problem that intellisense is displayed twice.
Confirmed with following.

```
.
└── parent
    ├── .graphqlrc.json
    ├── alpha
    │   ├── alpha.graphql
    │   └── foo.tsx
    ├── beta
    │   ├── beta.graphql
    │   └── foo.tsx
    └── shared
        └── shared.graphql
```

`parent/.graphqlrc.json`
```json
{
  "projects": {
    "alpha": {
      "schema": ["alpha/**", "shared/**"]
    },
    "beta": {
      "schema": ["beta/**", "shared/**"]
    }
  }
}
```

```graphql
# shared.graphql
type Shared {
  shared: Boolean
}

# alpha.graphql
type Query {
  sharedAlpha: Shared
}

# beta.graphql
type Query {
  sharedBeta: Shared
  betaAnother: BetaType
}

type BetaType {
  foo: Boolean
}
```

on `alpha/foo.tsx`

image

on `beta/foo.tsx`

image

Let me see if I can fix it.

**Comment by Foo-x** [comment]
Fixed the problem above.
image
image

But there is yet another problem that if there is a GraphQL config file outside the nested folders, the inner folders always belong to the outside folder as a project.

```
.
└── parent
    ├── .graphqlrc.json
    ├── alpha
    │   ├── alpha.graphql
    │   └── foo.tsx
    ├── beta
    │   ├── beta.graphql
    │   └── foo.tsx
    ├── gamma
    │   ├── .graphqlrc.json
    │   ├── foo.tsx
    │   └── gamma.graphql
    └── shared
        └── shared.graphql
```

`parent/.graphqlrc.json`

```json
{
  "projects": {
    "alpha": {
      "schema": ["alpha/**", "shared/**"]
    },
    "beta": {
      "schema": ["beta/**", "shared/**"]
    }
  }
}
```

on `gamma/foo.tsx`

image

If you don't add `parent` to your workspace, there is no problem.

image

I think the server needs to be implemented to solve this problem.

**Comment by acao** [comment]
Yes @Foo-x this is one of the known bugs in the LSP server I was referring to. You are free to try and fix it in the server in this PR, or we can fix it later. What you’ve implemented works great!

**Comment by Foo-x** [comment]
Ok, I'll try it!

**Comment by acao** [comment]
@Foo-x this is amazing work! Would you like to pair? I have a few more relatively simple optimizations we can add. Are you on the discord?

**Comment by acao** [comment]
I would also love to explain some problematic architecture and hacks of mine that you're inevitably stumbling into 😆 , and how I envision this server _could_ be architected better. For example, as you can tell, I put way too much logic in `MessageProcessor` that should be in `GraphQLCache` or `GraphQLLanguageService`. Many lsp packages that use `vscode-languageserver` use a `TextDocument` class, though really we could just enhance `GraphQLCache` and update/invalidate entries on the proper workspace events, and remove the parsing of huge globs of files by `GraphQLCache`, and problem solved!

more coming...

- I think this will be the most signficant perf optimization for the server overall: the server was originally pre-LSP if you can't tell, so we really haven't adopted the proper client listener model yet. getting rid of the server's usage of `watchman` was the first step. But I never properly implemented the client listener model fully, and I didn't get rid of some huge glob loads that happy on many events (in `GraphQLCache`)

**Comment by acao** [comment]
@Foo-x also, another huge and relatively low-hanging optimization that leads to deleting a bunch of code is rewriting `_updateGraphQLConfig()` to only accept a specific project from the file. Then the logic needs to move from a single init to tracking which projects have "pre-initialized" or not.

**Comment by acao** [comment]
I think it's possible if we do this, or something you feel is even more optimal, #2459 we can hopefully close as well, which is terrible, vscode-crashing perf issues on multi-project caused by this method!

**Comment by Foo-x** [comment]
I have never used discord, but I just installed it and joined the channel!
My account is `foo-x#1064`.
I can use it if needed.

And thank you for your very thorough explanation!
I understand a little better what needs to be fixed.
But I'm new to this repo, so I don't understand the architecture yet.
I'll take a look at it.
Then, I'll try to cache only specific project from the file.

**Comment by acao** [comment]
@Foo-x if you want to add that optimisation, then sure! I will reach out on discord

File: packages/graphql-language-service-server/src/MessageProcessor.ts [hunk 3]
```diff
@@ -89,58 +89,48 @@ function toPosition(position: VscodePosition): IPosition {
   return new Position(position.line, position.character);
 }
 
-export class MessageProcessor {
+class WorkspaceMessageProcessor {
```
Comment (acao) [inline packages/graphql-language-service-server/src/MessageProcessor.ts:92]: Brilliant! An elegant solution

Comment (acao) [reply packages/graphql-language-service-server/src/MessageProcessor.ts:92]: Would you be interested in separating the `WorkspaceMessageProcessor` into a separate file? the less cognitive overhead, the more useful it is as a reference implementation, the more accessible it is to contributors!

Comment (Foo-x) [reply packages/graphql-language-service-server/src/MessageProcessor.ts]: Thank you for your review!
Separated to new file.

**Review by acao — APPROVED** [review]
Everything here looks good to go! Thank you for helping with such a comprehensive PR!
Looking forward to more collaboration 🚀

**Comment by acao** [comment]
@Foo-x ok I hate to do this, it's my fault really. I merged too quickly. 

What I would like to do is see if you could re-push your branch if it's still on your local so we can revert this PR. Then I will merge #2573 and release, then merge a re-creation of this PR for the pre-release

**Comment by Foo-x** [comment]
@acao 
Of course, you can revert my PR!
I have created new PR that revert this.
https://github.com/graphql/graphiql/pull/2612

But how `.changeset/thick-years-sort.md` reverted since it has been changed on latter commits?
I simply deleted it on the PR above.
If there is another way, please let me know.

**Pull request merged**